src/Corax/Querying/Matches/SortingMatches/NumericalMaxHeapSorter.cs

```diff
@@ -0,0 +1,303 @@
+        {
+            Swap(parent, i);
+            i = parent;
+            parent = Parent(i);
+        }
+    }
+
+    private void MaxHeapify(int i)
+    {
+        Heapify:
+        int largest = i;
+        int leftChild = LeftChild(i);
+        int rightChild = RightChild(i);
+
+        // left child > current
+        if (leftChild < _heapSize)
+        {
+            var cmp = Compare(_terms[leftChild], _documents[leftChild], _terms[i], _documents[i]);
+
+            largest = cmp > 0
+                ? leftChild
+                : i;
+        }
+
+        // right child > largest
+        if (rightChild < _heapSize)
+        {
+            var cmp = Compare(_terms[rightChild], _documents[rightChild], _terms[largest], _documents[largest]);
+
+            if (cmp > 0)
+                largest = rightChild;
+        }
+
+        if (largest != i)
+        {
+            Swap(largest, i);
+
+            // Instead going recursively just jump to the beginning
+            i = largest;
+            goto Heapify;
+        }
+    }
+
+    private void ReplaceMax(TTermType term, int documentId)
+    {
+        _terms[0] = term;
+        _documents[0] = documentId;
+        MaxHeapify(0);
+    }
+
+    private void RemoveMax()
+    {
+        _terms[0] = _terms[_heapSize - 1];
+        _documents[0] = _documents[_heapSize - 1];
+        _heapSize--;
+
+        MaxHeapify(0);
+    }
+
+    public void Fill(Span<long> batchResults, ref ContextBoundNativeList<long> results, ref ContextBoundNativeList<float> scoreDestination, Span<float> scores)
+    {
+        ValidateMaxHeapStructure();
+        var start = results.Count;
+        results.EnsureCapacityFor(_heapSize);
+        int documentsToReturn = _heapSize;
+
+        var exposeScores = scoreDestination.HasContext && scores.Length != 0;
+        if (exposeScores)
+            scoreDestination.EnsureCapacityFor(_heapSize);
+        
+        while (documentsToReturn > 0)
+        {
+            results.AddUnsafe(batchResults[_documents[0]]);
+            
+            if (exposeScores)
+                scoreDestination.AddUnsafe(scores[_documents[0]]);
+            
+            RemoveMax();
+            documentsToReturn--;
+        }
+
+        Debug.Assert(_heapSize == 0, "_heapSize == 0");
+        // It's easier for us to operate on a MaxHeap, so we get the elements in descending order (locally).
+        // However, instead of rebuilding the heap as a MinHeap, let's add everything from the heap and call Reverse from the Span extension,
+        // which is a vectorized operation.
+        results.ToSpan().Slice(start).Reverse();
+    }
+
+    public void FillWithTerms(Span<long> batchResults, ref ContextBoundNativeList<long> results, ref ContextBoundNativeList<TTermType> terms, ref ContextBoundNativeList<float> scoreDestination, Span<float> scores)
+    {
+        ValidateMaxHeapStructure();
```

Comment:
Include documentation on what we are doing and why.

Reply:
done